artifactory/utils/container/containermanager.go

```diff
@@ -0,0 +1,350 @@
+	return exec.Command("sh", "-c", cmd)
+}
+
+func (loginCmd *LoginCmd) GetEnv() map[string]string {
+	return map[string]string{"CONTAINER_MANAGER_PASS": loginCmd.Password}
+}
+
+func (loginCmd *LoginCmd) GetStdWriter() io.WriteCloser {
+	return nil
+}
+
+func (loginCmd *LoginCmd) GetErrWriter() io.WriteCloser {
+	return nil
+}
+
+// Image pull command
+type pullCmd struct {
+	image            *Image
+	containerManager ContainerManagerType
+}
+
+func (pullCmd *pullCmd) GetCmd() *exec.Cmd {
+	var cmd []string
+	cmd = append(cmd, "pull")
+	cmd = append(cmd, pullCmd.image.tag)
+	return exec.Command(pullCmd.containerManager.String(), cmd[:]...)
+}
+
+func (pullCmd *pullCmd) GetEnv() map[string]string {
+	return map[string]string{}
+}
+
+func (pullCmd *pullCmd) GetStdWriter() io.WriteCloser {
+	return nil
+}
+
+func (pullCmd *pullCmd) GetErrWriter() io.WriteCloser {
+	return nil
+}
+
+// First will try to login assuming a proxy-less tag (e.g. "registry-address/docker-repo/image:ver").
+// If fails, we will try assuming a reverse proxy tag (e.g. "registry-address-docker-repo/image:ver").
+func ContainerManagerLogin(imageTag string, config *ContainerManagerLoginConfig, containerManager ContainerManagerType) error {
+	imageRegistry, err := ResolveRegistryFromTag(imageTag)
+	if err != nil {
+		return err
+	}
+	username := config.ArtifactoryDetails.User
+	password := config.ArtifactoryDetails.Password
+	// If access-token exists, perform login with it.
+	if config.ArtifactoryDetails.AccessToken != "" {
+		log.Debug("Using access-token details in " + containerManager.String() + "-login command.")
+		username, err = auth.ExtractUsernameFromAccessToken(config.ArtifactoryDetails.AccessToken)
+		if err != nil {
+			return err
+		}
+		password = config.ArtifactoryDetails.AccessToken
+	}
+	// Perform login.
+	cmd := &LoginCmd{ContainerRegistry: imageRegistry, Username: username, Password: password, containerManager: containerManager}
+	err = gofrogcmd.RunCmd(cmd)
+	if exitCode := coreutils.GetExitCode(err, 0, 0, false); exitCode == coreutils.ExitCodeNoError {
+		// Login succeeded
+		return nil
+	}
+	log.Debug(containerManager.String()+" login while assuming proxy-less failed:", err)
+	indexOfSlash := strings.Index(imageRegistry, "/")
+	if indexOfSlash < 0 {
+		return errorutils.CheckError(errors.New(fmt.Sprintf(LoginFailureMessage, containerManager.String(), imageRegistry, containerManager.String())))
+	}
+	cmd = &LoginCmd{ContainerRegistry: imageRegistry[:indexOfSlash], Username: config.ArtifactoryDetails.User, Password: config.ArtifactoryDetails.Password}
+	err = gofrogcmd.RunCmd(cmd)
+	if err != nil {
+		// Login failed for both attempts
+		return errorutils.CheckError(errors.New(fmt.Sprintf(LoginFailureMessage,
+			containerManager.String(), fmt.Sprintf("%s, %s", imageRegistry, imageRegistry[:indexOfSlash]), containerManager.String()) + " " + err.Error()))
+	}
+	// Login succeeded
+	return nil
+}
+
+// Version command
+// Docker only
```

Comment:
Let'e clerify here what "Docker only" means.

Reply:
`Docker-client provides an API for interacting with the Docker daemon. This cmd should be used for docker client only.`